### PR TITLE
TypedArrays: fixed find and findIndex callback param obj type.

### LINKED ISSUE #18425: TypedArrays: type thisArg correctly and other issues
For example, UInt8Array.find

```ts
// currently:
find(
    predicate: (value: number, index: number, obj: Array<number>) => boolean, 
    thisArg?: any): number | undefined;
// better:
find<Z = undefined>(
    predicate: (this: Z, value: number, index: number, obj: this) => boolean,
    thisArg?: Z): number | undefined;
```

All the variants have the same issues. It seems like it would be a good idea to refactor them to:
```ts
interface TypedArray<ArrType> {
// find etc
}
interface UInt8Array extends TypedArray<UInt8Array> {}
interface UInt16Array extends TypedArray<UInt16Array> {}
// etc
```

### PR DESCRIPTION
**Pull request by NaridaL**
Fixes #18425.

### CODE REVIEW

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -2644,7 +2644,7 @@ interface Uint16Array {
       * @param thisArg If provided, it will be used as the this value for each invocation of
       * predicate. If it is not provided, undefined is used instead.
       */
-    find(predicate: (value: number, index: number, obj: Array<number>) => boolean, thisArg?: any): number | undefined;
+    find(predicate: (value: number, index: number, obj: Int16Array) => boolean, thisArg?: any): number | undefined;
```
Comment (mhegazy) [inline src/lib/es5.d.ts]: Uint16Array ?

File: src/lib/es5.d.ts [hunk 2]
```diff
@@ -2655,7 +2655,7 @@ interface Uint16Array {
       * @param thisArg If provided, it will be used as the this value for each invocation of
       * predicate. If it is not provided, undefined is used instead.
       */
-    findIndex(predicate: (value: number, index: number, obj: Array<number>) => boolean, thisArg?: any): number;
+    findIndex(predicate: (value: number, index: number, obj: Int16Array) => boolean, thisArg?: any): number;
```
Comment (mhegazy) [inline src/lib/es5.d.ts]: Uint16Array

File: src/lib/es5.d.ts [hunk 3]
```diff
@@ -3178,7 +3178,7 @@ interface Uint32Array {
       * @param thisArg If provided, it will be used as the this value for each invocation of
       * predicate. If it is not provided, undefined is used instead.
       */
-    find(predicate: (value: number, index: number, obj: Array<number>) => boolean, thisArg?: any): number | undefined;
+    find(predicate: (value: number, index: number, obj: Int32Array) => boolean, thisArg?: any): number | undefined;
```
Comment (mhegazy) [inline src/lib/es5.d.ts]: Uint32Array

File: src/lib/es5.d.ts [hunk 4]
```diff
@@ -3189,7 +3189,7 @@ interface Uint32Array {
       * @param thisArg If provided, it will be used as the this value for each invocation of
       * predicate. If it is not provided, undefined is used instead.
       */
-    findIndex(predicate: (value: number, index: number, obj: Array<number>) => boolean, thisArg?: any): number;
+    findIndex(predicate: (value: number, index: number, obj: Int32Array) => boolean, thisArg?: any): number;
```
Comment (mhegazy) [inline src/lib/es5.d.ts]: Uint32Array

**Comment by NaridaL** [comment]
Yeah... sorry about that.

BTW, I tried it with the separate [TypedArray interface](https://github.com/NaridaL/TypeScript/blob/typedarray-dry/src/lib/es5.d.ts#L1777), and it seems to be an improvement...

previous:
```
PS C:\Users\aval\tsdev\TypeScript> node built\local\tsc.js --diagnostics --p src\compiler\tsconfig.json
Files:            36
Lines:         93186
Nodes:        445312
Identifiers:  168285
Symbols:       71668
Types:         32088
Memory used: 371617K
I/O read:      0.02s
I/O write:     0.17s
Parse time:    0.71s
Bind time:     0.62s
Check time:    2.90s
Emit time:     2.06s
Total time:    6.29s
```

now:
```
PS C:\Users\aval\tsdev\TypeScript> node built\local\tsc.js --diagnostics --p src\compiler\tsconfig.json
Files:            36
Lines:         91325
Nodes:        440475
Identifiers:  166994
Symbols:       70992
Types:         31941
Memory used: 350637K
I/O read:      0.03s
I/O write:     0.03s
Parse time:    0.69s
Bind time:     0.58s
Check time:    2.84s
Emit time:     1.88s
Total time:    5.98s
```

... not sure if those statistics are the only relevant thing though.

**Comment by NaridaL** [comment]
Oh yeah, the issues with U/Intarray should be fixed now, because it's not immediately obvious from the code reviews^^.

**Comment by mhegazy** [comment]
We should be moving them out of es5.d.ts anyways. https://github.com/Microsoft/TypeScript/pull/15780 should be in first though.

**Pull request merged**